[RAG] Stage - 3: Embeddings Generation

mulla028 opened the pull request:
## Description

This is **stage 3** of #868. We are adding the capability of generation and storage vector embeddings for document chunks. It introduces modular embedding provider archtecture, and supports:
* OpenAI's **text-embedding-3-small** API
* Local **tensorflow.js** alternative

[ChatCraftFile](https://github.com/tarasglek/chatcraft.org/blob/mulla028/868-3/src/lib/ChatCraftFile.ts) has been extended with methods to:

- generate embeddings
- store embeddings
- manage embeddings

Integrated embedding generation with the [use-file-import](https://github.com/tarasglek/chatcraft.org/blob/mulla028/868-3/src/hooks/use-file-import.tsx) to **automatically create embeddings after chunking**. 

**New settings added** to control:
- the embedding provider
- batch size
- automatic generation preference

> [!IMPORTANT]
> UI for the embedding preferences required! Therefore, we must land this PR with the updated Settings UI...

## Test It

>[!TIP]
> You may want to test it. In order to do so follow the steps below!

1. Open `CloudFlare` deployment below
2. Upload File **>=300KB**
3. Go to **DevTools**
4. `Application` → `Storage` → `IndexedDB` → `ChatCraftDatabase` → `files`
5. Wuolah! You can see the generated embeddings! (I HOPE :D) 

## Screenshot

<img width="1470" alt="image" src="https://github.com/user-attachments/assets/36fd071c-6a07-48f7-9a54-6918ab954045" />


#### Question 

Maybe we could reduce the size of the minimum chunking size from 300KB → 100 KB. Therefore, the minimum character per chunk is from 1000 → 300.

mulla028 committed:
TensorFlow.js and Universal Sentence Encoder for embeddings added

mulla028 committed:
Embedding provider interface for vector generation

mulla028 committed:
OpenAI text-embedding-3-small provider implemented

mulla028 committed:
TensorFlow-based local embedding provider added

mulla028 committed:
Embedding provider utilities implemented

mulla028 committed:
Embedding generation methods added

mulla028 committed:
Embedding settings to application config added

mulla028 committed:
File generates embeddings after it has been chunked

mulla028 commented:
Right now I am implementing the UI for RAG settings, and I am trying to figure out what batch size limit for the `TensorFlow`. According to some research, it depends on some kind of browser **WebGL**, I don't really know what's this. It says `50-100` chunks is the potential maximum, but `20-30` is a safer option. Just wanna know which values add to the slider. @humphd, maybe you know the answer.

mulla028 committed:
UI for Rag embedding settings implemented

mulla028 commented:
Since now, by default we do not generate embeddings

mulla028 committed:
Erased unsed variable

mulla028 commented:
By the way, using the openai option generates embeddings within seconds for those files that we were trying to upload within the lecture

mulla028 commented:
By increasing the tensorflow batch size from 20 to 40 chunks, it takes 5 seconds to generate embeddings for 11.3MB PDF file

mulla028 committed:
Updated embedding generation logic , using fetch vs. openai client, since error regarding API KEY revation

mulla028 committed:
autogeneration setting checks for chunk generation as well

humphd commented on src/lib/embeddings/OpenAIEmbedding.ts:
```diff
@@ -31,13 +30,27 @@ export class OpenAIEmbeddingProvider implements EmbeddingProvider {
-        input: texts,
-        encoding_format: "float",
+      const settings = getSettings();
+      console.log(`Current API KEY is: ${settings.currentProvider.apiKey}`);
```
Don't do this

humphd commented on src/lib/embeddings/OpenAIEmbedding.ts:
```diff
@@ -31,13 +30,27 @@ export class OpenAIEmbeddingProvider implements EmbeddingProvider {
-        encoding_format: "float",
+      const settings = getSettings();
+      console.log(`Current API KEY is: ${settings.currentProvider.apiKey}`);
+      const response = await fetch(`${this.url}`, {
```
Why can't use use the OpenAI client here?

mulla028 replied:
Is it safe to use fetch here? I used it instead of the OpenAI client because it helped me avoid this error:

> You have to set a flag to do it in a browser:
>
>Why is this dangerous?
Enabling the dangerouslyAllowBrowser option can be dangerous because it exposes your secret API credentials in the client-side code. Web browsers are inherently less secure than server environments, any user with access to the browser can potentially inspect, extract, and misuse these credentials. This could lead to unauthorized access using your credentials and potentially compromise sensitive data or functionality.
>
>When might this not be dangerous?
In certain scenarios where enabling browser support might not pose significant risks:
>
>Internal Tools: If the application is used solely within a controlled internal environment where the users are trusted, the risk of credential exposure can be mitigated.
Public APIs with Limited Scope: If your API has very limited scope and the exposed credentials do not grant access to sensitive data or critical operations, the potential impact of exposure is reduced.
Development or debugging purpose: Enabling this feature temporarily might be acceptable, provided the credentials are short-lived, aren't also used in production environments, or are frequently rotated.

humphd replied:
Using the OpenAI client is better--it's a wrapper around `fetch()` with extras implemented for you.

mulla028 replied:
Oh, that makes sense. I will revert it to the OpenAI client usage but gotta find a way to avoid that error.

mulla028 replied:
Alright, I reverted it and added a flag to the OpenAI client `dangerouslyAllowBrowser: true`, hope this is a way, and I hope chatcraft is safe enough to use that flag.

humphd replied:
It's not that it's dangerous as such.  It's more that they want you to understand that using an API key in browsers is obviously dangerous in general.  They want you to think through what you're doing and opt into using it.

But it's the same as you passing it to the headers of the `fetch()` call.

mulla028 replied:
Thank you for the explanation! It makes sense now

mulla028 commented:
I may confirm that now embeddings are properly generated for both strategies: `tensorflow` and `openai`. All settings work as expected, but I had to change the way I use the openai method since the client version wasn't allowing it for the reason that it has an ERROR: **Seams like you use browser-like etc... API KEY MAY BE REVEALED**, and I had to use fetch. The last thing is that I am going to try to use `Worker` for the parallel threading and make a progress bar update while generating both chunks and embeddings. After we finish landing the RAG I will have a follow-up for  UI improvement during **file import**.

mulla028 commented:
Do I remember correctly that we need to make huge files, for example, >5MB to be generated using openai only, or to make an alert for the user that he could use Openai, so he isn't forced and still has a choice

humphd commented:
You have to set a flag to do it in a browser:

>Why is this dangerous?
>Enabling the dangerouslyAllowBrowser option can be dangerous because it exposes your secret API credentials in the client-side code. Web browsers are inherently less secure than server environments, any user with access to the browser can potentially inspect, extract, and misuse these credentials. This could lead to unauthorized access using your credentials and potentially compromise sensitive data or functionality.
>
>When might this not be dangerous?
>In certain scenarios where enabling browser support might not pose significant risks:
>
>Internal Tools: If the application is used solely within a controlled internal environment where the users are trusted, the risk of credential exposure can be mitigated.
Public APIs with Limited Scope: If your API has very limited scope and the exposed credentials do not grant access to sensitive data or critical operations, the potential impact of exposure is reduced.
>Development or debugging purpose: Enabling this feature temporarily might be acceptable, provided the credentials are short-lived, aren't also used in production environments, or are frequently rotated.

mulla028 commented:
You wanna see the following changes: 
* revert to openai usage vs. fetch
* add that flag ?

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,68 @@
+          value={settings.embeddingBatchSize}
+          onChange={(value) => setSettings({ ...settings, embeddingBatchSize: value })}
+          min={20}
+          max={80}
```
Why 80?  OpenAI can do way more.  Is Tensorflow.js limited?

mulla028 replied:
Tbh, I don't really understand their limits, but now I am testing different sizes for both providers. If you know, could you explain to me how I learn about the TensorFlow limits since I don't use the tokens and every user will have their own backend used(if there is a way to learn about it without the testing)?

mulla028 replied:
Even 1000 chunks work and it's rapid 😆

humphd replied:
OpenAI can use thousands, and tensorflow.js will be limited by available RAM.  So I think the range we provide here needs to be built-into the embedding provider, since the two differ so much.

mulla028 replied:
Do we still want to provide the choice of batch size in the settings? Or just try to set the default value and max value based on the provider?

humphd replied:
I think we should, but when using tensorflow.js, we'd use 16, 32, 64, etc. and for OpenAI we can do a lot more.

mulla028 replied:
Let's say that the limit for tensorflow.js is 256, I tried a huge pdf that generated 250 chunks and it worked well with a batch size of 220 ish, but the default value is going to be 128. For OpenAI I am going to set the default value as 2500  and the max as 4000.

humphd replied:
I think 4000 is low, but let's try and adjust in follow-ups.

humphd commented on src/hooks/use-file-import.tsx:
```diff
@@ -217,12 +217,22 @@ export function useFileImport({ chat, onImageImport }: UseFileImportOptions) {
       await chat.addFile(chatCraftFile);
 
       // Generate chunks for the file if its size > 300
```
This comment doesn't totally make sense with the code that follows.  You're not looking at the file size here, only whether there are already chunks and what the preferences say.

mulla028 replied:
Forgot to delete it. I am cleaning it up

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -180,6 +182,28 @@ export class ChatCraftFile {
+      return undefined;
+    }
+
+    return this.chunks[0].metadata.embeddingDimensions as number;
```
The fact that here and on line 193 you have to get the first chunk in order to get the metadata you need makes me think that we should lift this info out of the chunks and store it on its own object.

mulla028 replied:
Can we store it inside `this.metadata` instead of `this.chunks[0].metadata`?

humphd replied:
That makes sense to me.

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -205,6 +229,64 @@ export class ChatCraftFile {
+  hasEmbeddings(): boolean {
+    return (
+      this.hasChunks() &&
+      this.chunks!.every((chunk) => Array.isArray(chunk.embeddings) && chunk.embeddings.length > 0)
```
Don't use `!` in TypeScript like this, you're turning off the type checking.

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -205,6 +229,64 @@ export class ChatCraftFile {
+
+    const BATCH_SIZE = settings.embeddingBatchSize || 20;
+
+    for (let i = 0; i < chunks.length; i += BATCH_SIZE) {
```
This code can move to a worker thread in a follow-up.

humphd commented on src/lib/embeddings/TensorflowEmbedding.ts:
```diff
@@ -0,0 +1,78 @@
+   * Load the model if it hasn't been loaded yet
+   */
+  private async loadModelIfNeeded(): Promise<void> {
+    if (this.model) return;
```
Always brace your if

humphd commented on src/lib/embeddings/TensorflowEmbedding.ts:
```diff
@@ -0,0 +1,78 @@
+        try {
+          console.log("Loading Universal Sentence Encoder model...");
+
+          await import("@tensorflow/tfjs");
```
Question: will this load the best backend (e.g., GPU or WebGL versions) by default?

mulla028 replied:
I wouldn't say it loads the best by default, but loads the best **available** by default

mulla028 replied:
TensorFlow has a setBackend() method, where I could specify one. Do you have a preference?

humphd replied:
We can't know what the user will have (e.g., I might have a GPU, another user doesn't).  The goal would be that we load the most capable backend at runtime.

This can happen in a follow-up, don't let it block things here.  I was just curious.

humphd commented on src/lib/embeddings/TensorflowEmbedding.ts:
```diff
@@ -0,0 +1,78 @@
+
+      const arrays = await embeddings.array();
+
+      embeddings.dispose();
```
This needs to go in a `finally` probably, so we don't leak

humphd reviewed (changes requested):
This is looking really good.  Some issues to sort out now, and some more for follow-ups.

I love that you also added some UI!

mulla028 committed:
Deleted debug stmt

mulla028 committed:
Deleted debug stmt

mulla028 committed:
braced return stmt

mulla028 committed:
lifted up embeddings metadata and fixed setMetadata method

mulla028 commented:
The `setMetadata()` method wasn't working properly tho. It wasn't updating the db for some reason.

mulla028 committed:
Typo

mulla028 committed:
dispose moved to finally

mulla028 committed:
Reverted to use openai client, added flag to avoid the error regarding apiKey leak

humphd commented:
It looks like tensorflow.js will automatically try to load the best backend for your browser/hardware.

mulla028 commented:
`await import("@tensorflow/tfjs");` it automatically loads the best backend available for your browser/hardware

mulla028 committed:
Attribute added to embedding provider interface, now tfjs 256 max batches and openai 4000

mulla028 commented:
Tbh, I don't like my UI solution for the batch size adjustment logic it looks ugly, and I believe there should be a smarter way.

mulla028 commented:
Despite the UI solution, I think this draft PR is ready to become at least an actual PR. Do we need to switch providers to OpenAI if huge files are uploaded? Or we could focus on the main core of landing RAG, and then implement it as a follow-up leaving it as an experimental feature

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+        <Slider
+          value={settings.embeddingMaxBatchSize}
+          onChange={(value) => setSettings({ ...settings, embeddingMaxBatchSize: value })}
+          min={settings.embeddingProvider === "openai" ? 500 : 16}
```
Move these values into the providers as `minChunkBatchSize` and `maxChunkBatchSize` properties

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+            setSettings({
+              ...settings,
+              embeddingProvider: value as EmbeddingProviderType,
+              embeddingMaxBatchSize: value === "openai" ? 4000 : 256,
```
Move this into the provider as `defaultChunkBatchSize`

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+        >
+          <Stack>
+            <Radio value="openai">Open-AI (Requires API KEY!)</Radio>
+            <Radio value="tensorflow">Tensorflow</Radio>
```
"Generate in browser (tensorflow.js)"

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+          isDisabled={!settings.autogenerateEmbeddings}
+        >
+          <Stack>
+            <Radio value="openai">Open-AI (Requires API KEY!)</Radio>
```
"Use OpenAI (requires API Key)"

humphd commented on src/components/Preferences/RagSettings.tsx:
```diff
@@ -0,0 +1,72 @@
+          </SliderTrack>
+          <SliderThumb />
+        </Slider>
+        <FormErrorMessage>Maximum batch size must be between 20 and 80 chunks</FormErrorMessage>
```
This needs to match what the provider supports

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -93,6 +95,7 @@ export class ChatCraftFile {
     input: File | ChatCraftFileOptions,
     additionalOptions: Partial<ChatCraftFileOptions>
   ): Promise<ChatCraftFile> {
+    console.log("findOrCreate invoked");
```
Remove this.

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -105,6 +108,8 @@ export class ChatCraftFile {
       };
 
+      console.log("Metadata that is getting used", options.metadata);
+      console.log("Additional Metadata: ", additionalOptions);
```
Remove these

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -180,14 +185,37 @@ export class ChatCraftFile {
+  /**
+   * Get the embedding provider used for this file's chunks
+   */
+  getEmbeddingProviderId(): string | undefined {
```
Why not `getEmbeddingsProvider()` and return an object that has all the info for the provider?

mulla028 replied:
I agree

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -180,14 +185,37 @@ export class ChatCraftFile {
+  /**
+   * Get embedding dimensions
+   */
+  getEmbeddingDimensions(): number | undefined {
```
Use "Embeddings" (plural) everywhere

humphd commented on src/lib/ChatCraftFile.ts:
```diff
@@ -205,6 +233,60 @@ export class ChatCraftFile {
+      return false;
+    }
+
+    return this.chunks?.some((chunk) => chunk.embeddings.length > 0);
```
Use `.every()`

humphd commented on src/lib/embeddings/index.ts:
```diff
@@ -0,0 +1,53 @@
+    settings.embeddingMaxBatchSize = providers[type].maxBatchSize;
+  }
+
+  return providers[type]!;
```
Don't use `!`

mulla028 committed:
Added config for better settings handling

mulla028 committed:
slightly changed logic to use the config

mulla028 committed:
Renamed to use plural, added getEmbeddingsProvider and removed other getters

mulla028 committed:
plural

mulla028 commented:
Alright, now we have a single embeddings metadata getter in `ChatCraftFile.ts`

`src/lib/embeddings`:
* Added `minBatchSize`, `maxBatchSize` and `defaultBatchSize`
* Created configs
 
 `src/lib/settings.ts`:
 * renamed properties
 
 Based on the change request made a clean up.
 
 I don't know if you like this config approach, I just couldn't come up with the better one. Let me know if I am missing something or if you think could be implemented in another way

mulla028 commented:
I know that we are all taking a break after this hard term, and I hope that we are still implementing this feature. However, I am inactive till the May, 3. Still waiting for the feedback, let me know when I have to expect it (without pushing anyone, probably you are also taking a break now.) Just let me know when to expect the feedback, since I haven't received anything yet, according to the latest updates. Have a good break :)

mulla028 commented:
@humphd Could you review this pull-request, or do we have to figure [it](https://github.com/tarasglek/chatcraft.org/issues/874) out first?

humphd commented:
@mulla028 it's on my list, and getting this stuff fully integrated is something I want to do.  But I'm currently tied up with other projects.  TODO...

tarasglek commented:
I'm concerned about UX after this change. Now it's completely not clear if file is gonna get RAGed or wholly passed to LLM.

mulla028 commented:
> I'm concerned about UX after this change. Now it's completely not clear if file is gonna get RAGed or wholly passed to LLM.

We have already discussed it with @humphd. We came up with the idea of experimental feature until we land all the stages. Now in the settings you may pick whether you want embedding to be generated(for the further stages) or not. By default its off

mulla028 commented:
![image](https://github.com/user-attachments/assets/8ee10b7e-9099-4124-8fad-72dfc4c974f6)
